### PR TITLE
refactor: onSnapshot method in Watch

### PR DESCRIPTION
**Pull request by thebrianchen**

### CODE REVIEW

File: dev/src/watch.ts [hunk 1]
```diff
@@ -765,70 +431,435 @@ abstract class Watch {
...
+
+    if (!this.hasPushed || appliedChanges.length > 0) {
+      logger(
+        'Watch.onSnapshot',
+        this.requestTag,
+        'Sending snapshot with %d changes and %d documents',
+        String(appliedChanges.length),
+        this.docTree.length
+      );
+      // Pass doctree in as a const, so we don't lose reference to 'this'.
```
Comment (schmidt-sebastian) [inline dev/src/watch.ts]: That's not really the reason we are doing this. We want to always refer to what `this.docTree` points to, and not to what the `docTree` in `this` points to. Maybe make this something like:

```
// We pass the current set of changes, even if `docTree` is modified later.
const currentState = this.docTree;
```

File: dev/src/watch.ts [hunk 2]
```diff
@@ -765,70 +431,435 @@ abstract class Watch {
...
+      );
+      this.hasPushed = true;
+    }
+
+    this.changeMap.clear();
+    this.resumeToken = nextResumeToken;
+  }
+
+  /**
+   * Applies a document delete to the document tree and the document
```
Comment (schmidt-sebastian) [inline dev/src/watch.ts]: Nit: Reflow the comments here and below.

File: dev/src/watch.ts [hunk 3]
```diff
@@ -318,393 +330,47 @@ abstract class Watch {
...
+      this.onError.toString() === EMPTY_FUNCTION.toString(),
+      'onError should not already be defined.'
+    );
+    assert(
+      this.docTree === undefined,
+      'docTree should not already be defined.'
+    );
+    this.onNext = onNext;
+    this.onError = onError;
+    this.docTree = rbtree(this.getComparator());
```
[the remark this answers is not in the document]
Comment (thebrianchen) [reply dev/src/watch.ts]: Initializing docTree in the constructor causes the tests to fail.

Initializing `QueryWatch` calls `super(firestore)` which goes into the `Watch` constructor. If `this.getComparator()` is called then, `query` and `comparator` are not defined.

File: dev/src/watch.ts [hunk 4]
```diff
@@ -765,70 +431,435 @@ abstract class Watch {
...
+   * Returns the current count of all documents, including the changes from
+   * the current changeMap.
+   */
+  private currentSize(): number {
+    const changes = this.extractChanges(Timestamp.now());
+    return this.docMap.size + changes.adds.length - changes.deletes.length;
+  }
+
+  /** Splits up document changes into removals, additions, and updates. */
+  private extractChanges(readTime: Timestamp): DocumentChangeSet {
```
[the remark this answers is not in the document]
Comment (thebrianchen) [reply dev/src/watch.ts]: `extractChanges` is similarly named in the [java sdk](https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/Watch.java#L437). 

Is it ok to diverge in naming, or should we keep them consistent?

[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply dev/src/watch.ts]: Let's keep it as is.

**Pull request merged**